### PR TITLE
More cleanup in advance of DB Partitioning Mode

### PR DESCRIPTION
**Pull request by jamesagnew**
This is part 4 of the database partitioning mode changes. This PR is a second set of changes that don't really introduce any new functionality, but:

* Do a bunch of cleanup of tests, including fixing some intermittents related to the logback extension
* Add a RequestDetails to a bunch of the MDM API, which will be needed for eventually partition-enabling these APIs.
* Removes the redundant RequestDetails object that is passed around QueryStack and the rest of the SQL generation methods (it's already available as a field in QueryStack so it doesn't need to be a method parameter too)
* Adds better generic types to a few places for better type safety

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 1]
```diff
@@ -821,6 +821,7 @@ private Condition createMissingPredicateForUnindexedMissingFields(
 
 		ICanMakeMissingParamPredicate innerQuery = PredicateBuilderFactory.createPredicateBuilderForParamType(
 				theParams.getParamType(), theParams.getSqlBuilder(), this);
+		assert innerQuery != null;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:824]: suggestion: replace with `Validate.validateNotNull();` possibly, as sonar will cry about asserts in production code.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/search/builder/SearchBuilderTest.java [hunk 2]
```diff
@@ -68,4 +72,58 @@ void testCalculateIndexUriIdentityHashesForResourceTypes_RevInclude_Null() {
...
+			List.of(
+				JpaPid.fromId(103L, 1)),
+			List.of(
+				JpaPid.fromId(200L, 2),
+				JpaPid.fromId(201L, 2),
+				JpaPid.fromId(202L, 2)),
+			List.of(
+				JpaPid.fromId(203L, 2)
+			)
+		);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/search/builder/SearchBuilderTest.java:126]: nice!

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4IndexStorageOptimizedTest.java [hunk 3]
```diff
@@ -79,6 +80,16 @@ public class FhirResourceDaoR4IndexStorageOptimizedTest extends BaseJpaR4Test {
...
 	private SearchConfig mySearchConfig;
 
+	@Override
+	@BeforeEach
+	public void before() throws Exception {
+		super.before();
+
+		// We rely on this interceptor being in place, and it should be unless some cheeky other test
+		// has removed it
+		assertEquals(1, myInterceptorRegistry.getAllRegisteredInterceptors().stream().filter(t -> t instanceof SearchParamValidatingInterceptor).count());
```
Comment (tadgh) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4IndexStorageOptimizedTest.java:90]: hehe

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/search/reindex/InstanceReindexServiceImplR5Test.java [hunk 4]
```diff
@@ -344,15 +345,16 @@ private static Parameters.ParametersParameterComponent findSingleIndex(Parameter
...
+		Validate.notBlank(theParamName, "theParamName must not be blank");
 		List<Parameters.ParametersParameterComponent> indexes = theResponse.getParameters(theSectionName);
 		assertThat(indexes).hasSize(1);
 
 		List<Parameters.ParametersParameterComponent> indexInstances = indexes
 			.get(0)
 			.getPart()
 			.stream()
-			.filter(t -> t.getName().equals(theParamName))
+			.filter(t -> theParamName.equals(t.getName()))
```
Comment (tadgh) [inline hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/search/reindex/InstanceReindexServiceImplR5Test.java:357]: nitpick: may as well use a method reference

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/expunge/PartitionAwareSupplier.java [hunk 5]
```diff
@@ -33,14 +33,16 @@ public class PartitionAwareSupplier {
...
 		myTransactionService = theTxService;
 		myRequestDetails = theRequestDetails;
 	}
 
 	@Nonnull
 	public <T> T supplyInPartitionedContext(Supplier<T> theResourcePersistentIdSupplier) {
-		return myTransactionService.withRequest(myRequestDetails).execute(tx -> theResourcePersistentIdSupplier.get());
+		T retVal =
+				myTransactionService.withRequest(myRequestDetails).execute(tx -> theResourcePersistentIdSupplier.get());
+		assert retVal != null;
```
Comment (tadgh) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/expunge/PartitionAwareSupplier.java:45]: nitpick: assert vs validate.XX

File: hapi-tinder-plugin/src/main/java/ca/uhn/fhir/tinder/TinderJpaRestServerMojo.java [hunk 6]
```diff
@@ -130,7 +130,7 @@ public void execute() throws MojoExecutionException, MojoFailureException {
 			baseResourceNames.removeAll(excludeResourceNames);
 		}
 
-		ourLog.info("Including the following resources: {}", baseResourceNames);
+		ourLog.debug("Including the following resources: {}", baseResourceNames);
```
Comment (tadgh) [inline hapi-tinder-plugin/src/main/java/ca/uhn/fhir/tinder/TinderJpaRestServerMojo.java:133]: 👍🏻

**Review by tadgh — APPROVED** [review]
Looks good to me, minor nitpicks.

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/util/SqlUtil.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+	/**
+	 * Non instantiable
+	 */
+	private SqlUtil() {
+		// nothing
+	}
+
+	@Nonnull
+	public static List<String> splitSqlFileIntoStatements(String theSql) {
```
Comment (michaelabuckley) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/util/SqlUtil.java:40]: minor: This is complicated enough to deserve a couple of tests.

**Review by michaelabuckley — APPROVED** [review]
Nice cleanup.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:824]: Makes sense. Done.

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/search/reindex/InstanceReindexServiceImplR5Test.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/search/reindex/InstanceReindexServiceImplR5Test.java:357]: I don't actually think that's possible here - the method needs an argument and it's an instance method so I don't believe there's any syntax for that?

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/expunge/PartitionAwareSupplier.java (hunk 5 above)
Comment (jamesagnew) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/expunge/PartitionAwareSupplier.java:45]: Changed

File: hapi-fhir-jpaserver-base/pom.xml [hunk 8]
```diff
@@ -413,6 +413,7 @@
 			<plugin>
 				<groupId>ca.uhn.hapi.fhir</groupId>
 				<artifactId>hapi-tinder-plugin</artifactId>
+				<version>${project.version}</version>
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/pom.xml:416]: Yeah, I guess you're right. Fixing this.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/partition/RequestPartitionHelperSvc.java [hunk 9]
```diff
@@ -132,7 +132,11 @@ public RequestPartitionId validateAndNormalizePartitionNames(RequestPartitionId
 			if (theRequestPartitionId.hasPartitionIds()) {
 				if (partition == null) {
 					Validate.isTrue(
-							theRequestPartitionId.getPartitionIds().get(i) == null,
+							theRequestPartitionId.getPartitionIds().get(i) == null
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/partition/RequestPartitionHelperSvc.java:135]: Good call. Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 10]
```diff
@@ -943,32 +944,27 @@ private Condition createPredicateFilter(
 			QueryStack theQueryStack3,
 			SearchFilterParser.BaseFilter theFilter,
 			String theResourceName,
-			RequestDetails theRequest,
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:946]: Definitely thought about it... And there may well be sense in it. My one major hesitation is that it becomes clunky (and easy to miss) in all the cases where you aren't actually serving an HTTP request.

Eg if you're writing a batch process or a scheduled job etc. that needs to make a DAO call, it's hard to miss the parameter that requires a SystemRequestDetails. But if we make it implicit then people will have to do calls like we do with `Roles.doAs(...)` in CDR, or we make the system assume that the absence of a RequestDetails in the threadlocal means it's a system request? (risky). I dunno, I can see it both ways. We should discuss this at the design meeting...

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FileUtil.java [hunk 11]
```diff
@@ -33,4 +41,17 @@ public static String formatFileSize(long theBytes) {
 		digitGroups = Math.min(digitGroups, UNITS.length - 1);
 		return new DecimalFormat("###0.#").format(theBytes / Math.pow(1024, digitGroups)) + " " + UNITS[digitGroups];
 	}
+
+	/**
+	 * Loads the contents of a File as a UTF-8 encoded string, and wraps any exceptions in an unchecked exception
+	 *
+	 * @throws InternalErrorException If any IOException occurs
+	 */
+	public static String loadFileAsString(File theFile) throws InternalErrorException {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FileUtil.java:50]: Yup, that makes a lot more sense then reinventing the wheel. Thanks!

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/util/SqlUtil.java (hunk 7 above)
Comment (jamesagnew) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/util/SqlUtil.java:40]: Yeah good call. Added (and found a bug!)

**Pull request merged**